### ISSUE TITLE
FreeBSD/arm64: POSIX/t/math.t fails with -Duselongdouble

### ISSUE THREAD

**Issue by eserte** [Perl/perl5#22463]
Module: POSIX

**Description**

During `make test` a `tan` test in`ext/POSIX/t/math.t` fails if perl was compiled with `-Duselongdouble`. Seems to happen only with FreeBSD/arm64, not with FreeBSD/amd64. Linux/arm64 is also not affected.

```
not ok 40 - tan(1) == -tan(-1)
#   Failed test 'tan(1) == -tan(-1)'
#   at ext/POSIX/t/math.t line 52.
#          got: 1.55740772465490223050697480745836
#     expected: 1.55740772465490223050697480745836
```

**Steps to Reproduce**

`./perl -Ilib ext/POSIX/t/math.t`

**Perl configuration**

```
Summary of my perl5 (revision 5 version 40 subversion 0) configuration:
   
  Platform:
    osname=freebsd
    osvers=14.1-stable
    archname=aarch64-freebsd-ld
    uname='freebsd cloud9.bbbike.org 14.1-stable freebsd 14.1-stable stable14-n268370-b37a6d41a046 generic arm64 '
    config_args='-ds -e -Dprefix=/usr/perl5.40.0D -Duselongdouble -Dcf_email=srezic@cpan.org -Doptimize=-O2 -pipe'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=define
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DNO_POSIX_2008_LOCALE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    optimize='-O2 -pipe'
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -DNO_POSIX_2008_LOCALE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='FreeBSD Clang 18.1.6 (https://github.com/llvm/llvm-project.git llvmorg-18.1.6-0-g1118c2e05e67)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=1
    ivtype='long'
    ivsize=8
    nvtype='long double'
    nvsize=16
    Off_t='off_t'
    lseeksize=8
    alignbytes=16
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags ='-Wl,-E  -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib /usr/lib/clang/18/lib
    libs=-lpthread -lgdbm -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags='-DPIC -fPIC'
    lddlflags='-shared  -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_LONG_DOUBLE
    USE_PERLIO
    USE_PERL_ATOF
  Built under freebsd
  Compiled at Aug  4 2024 05:20:26
  %ENV:
    PERLDOC="-MPod::Perldoc::ToTextOverstrike"
  @INC:
    lib
    /usr/perl5.40.0D/lib/site_perl/5.40.0/aarch64-freebsd-ld
    /usr/perl5.40.0D/lib/site_perl/5.40.0
    /usr/perl5.40.0D/lib/5.40.0/aarch64-freebsd-ld
    /usr/perl5.40.0D/lib/5.40.0
```

**Comment by jkeenan** [comment]
> Module: POSIX
> 
> **Description**
> 
> During `make test` a `tan` test in`ext/POSIX/t/math.t` fails if perl was compiled with `-Duselongdouble`. Seems to happen only with FreeBSD/arm64, not with FreeBSD/amd64. Linux/arm64 is also not affected.
> 
> ```
> not ok 40 - tan(1) == -tan(-1)
> #   Failed test 'tan(1) == -tan(-1)'
> #   at ext/POSIX/t/math.t line 52.
> #          got: 1.55740772465490223050697480745836
> #     expected: 1.55740772465490223050697480745836
> ```
> 

Is it your impression that this is a regression from a point in time when perl, built on this platform with those configuration arguments, passed this test?  If so, can you bisect?

**Comment by sisyphus** [comment]
The failing test assumes that tan(-1) is simply going to be the negation of tan(1) - ie tan(-1) = -tan(1).
In that case, the test inevitably passes because the condition tan(1) == -(-tan(1)) has to be true.

I suspect that in this particular failing case, tan(-1) is instead being calculated as tan((2 * pi) - 1).
This is (mathematically) equally valid, but in practice it's slightly less accurate because of the inexactness of pi.
There might be other scenarios that account for the discrepancy, but that was the only one I could think of.

I personally don't see a need to perform this test. Does it matter if tan(-1) is not being conducted as -tan(1) ?
Sure, that's the better and smarter approach, but if some underlying C system is going to do it differently, then I don't think our test suite should be registering a FAIL.
That it's being calculated to a satisfactory level of accuracy should be all that we're concerned about.

The math.t script contains:
```
...
between(1.55, tan(1), 1.56, 'tan(1)');
between(1.55, tan(1), 1.56, 'tan(-1)');
cmp_ok(tan(1), '==', -tan(-1), 'tan(1) == -tan(-1)');
...
```
The bottom of those 3 is the failing test being examined here.
The 2 tests above it are exactly the same (but with different names).
I think the bottom of those 2 tests was intended to be written as:
```
between(-1.55, tan(-1), -1.56, 'tan(-1)');
```
which is something that should be addressed, too.

For no particular reason, I calculated tan(-1) to 113-bit precision using both methods.
According to Math::MPFR:
tan(-1) => -1.55740772465490223050697480745836023
tan((2*pi)-1) => -1.55740772465490223050697480745836081

We see that both estimations round to the same 33-siginificant-digit value of 1.55740772465490223050697480745836, as the test suite reported.
Of course, I don't know exactly how the figures are being calculated in this build of perl.
It's a pity that Test::More displays "got" and "expected" NVs by doing (something like) "print $nv", instead of doing "printf '%a', $nv" - as the latter would display the exact values.

**Comment by sisyphus** [comment]
I'll prepare a PR that implements the following change to math.t:
```
$ diff -u math.t_orig math.t
--- math.t_orig 2024-08-05 20:21:20.763857600 +1000
+++ math.t      2024-08-05 20:43:29.575809900 +1000
@@ -48,8 +48,8 @@
 between(-1.18, sinh(-1), -1.17, 'sinh(-1)');
 is(tan(0), 0, "Basic tan(0) test");
 between(1.55, tan(1), 1.56, 'tan(1)');
-between(1.55, tan(1), 1.56, 'tan(-1)');
-cmp_ok(tan(1), '==', -tan(-1), 'tan(1) == -tan(-1)');
+between(-1.56, tan(-1), -1.55, 'tan(-1)');
+between(-1e-12, tan(1) + tan(-1), 1e-12, 'tan(1) + tan(-1) == (or closely approximates) 0');
 is(tanh(0), 0, "Basic tanh(0) test");
 between(0.76, tanh(1), 0.77, 'tanh(1)');
 between(-0.77, tanh(-1), -0.76, 'tanh(-1)');
```

**Comment by tonycoz** [comment]
Fixed by #22467
